Make loader strategy error message more verbose

jvllmr opened the pull request:
Fixes: #12398
This PR makes the error message on conflicting loader strategies more verbose to help with debugging

<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->
I changed the omitted error messages and the tests covering it.
### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

zzzeek commented on lib/sqlalchemy/orm/strategy_options.py:
```diff
@@ -1871,7 +1871,7 @@ def _reconcile(
 
         raise sa_exc.InvalidRequestError(
-            f"Loader strategies for {replacement.path} conflict"
+            f"Loader strategy replacement {replacement.path} is in conflict with existing strategy {existing.path}"
```
this is fine though I think you need to install pre-commit and have it run because this will fail pep8 for line length

jvllmr replied:
Thanks for the hint. I indeed forgot to lint my changes.

Note that pre-commit did not work for me because of a version conflict with black:
```
stderr:
    ERROR: Package 'black' requires a different Python: 3.8.19 not in '>=3.9'
Check the log at /home/jan/.cache/pre-commit/pre-commit.log
```

I used `tox -e lint` instead.

jvllmr committed:
more verbose loader strategy error
Fixes: #12398
this commit makes the error message on conflicting loader strategies more verbose to help with debugging

CaselIT commented on test/orm/test_deferred.py:
```diff
@@ -534,8 +534,10 @@ def go():
+            r"Loader strategy replacement "
+            r"ORM Path\[Mapper\[Order\(orders\)\] -> Order.user_id\] "
+            r"is in conflict with existing strategy "
+            r"ORM Path\[Mapper\[Order\(orders\)\] -> Order.user_id\]",
```
ok this doesn't seem to add much value, since the two sides are the same.

@zzzeek do we have also a way of getting what option is used in each part of the path?

jvllmr replied:
I thought bringing in `_LoadElement.strategy` might add more value to this, but I realised that it is also equal in a lot of cases.

zzzeek replied:
im not really sure, I haven't looked at this code in a long time.  I do know it's vastly simpler code than it was in 1.4 and earlier

jvllmr committed:
make changes actually useful

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
Merge branch 'main' into strategy-conflict-more-verbose

jvllmr committed:
remove unintended newlines

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

CaselIT commented:
thanks they do look a bit more informative

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 4de2e161924d2d61ee7ad8402315da754e7995f4 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 4de2e161924d2d61ee7ad8402315da754e7995f4: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

sqla-tester commented on lib/sqlalchemy/orm/strategy_options.py:
```diff
@@ -1865,9 +1865,25 @@ def _reconcile(
         )
 
+    def __str__(self) -> str:
+        args: dict[str, Any] = {
```
**Federico Caselli** (CaselIT) wrote:

we haven't really solved the comment that was mentioned in the pr, meaning the additional log doesn't really gives the user much more information

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

sqla-tester replied:
**Michael Bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

the change it fine implementation wise, I'm just not sure if it brings much value

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

well here's the "full" version.   what do you think of this one?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/strategy_options.py (line 1915): not crazy about this hardcoding here

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/strategy_options.py (line 1915): Done

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

nice work! left a couple of minor suggestions

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/strategy_options.py (line 104): we could maybe use _entity_cls_str hede?
* lib/sqlalchemy/orm/strategy_options.py (line 2455): ```suggestion
) -> Callable[[_FN], _FN]:

```

I don't thing we need the string here
* lib/sqlalchemy/orm/strategy_options.py (line 2456): let's use normal ascii characters 
```suggestion
    """Decorator that registers strategy key(s) -> function name in
```

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/strategy_options.py (line 2456): oops the robot showed his claw...

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

thanks for the review!  you totally caught a bunch of bad LLM decisions.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/strategy_options.py (line 104): Done
* lib/sqlalchemy/orm/strategy_options.py (line 2455): Done
* lib/sqlalchemy/orm/strategy_options.py (line 2456): Done

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/strategy_options.py (line 103): how come we aren't sure prop has a "key"?


also should we perhaps have this be a method on PathRegistry?   maybe not `__str__()` but perhaps `path_string()` ?

as part of this, consider existing `__str__()` on PathRegistry, possibly.  maybe it's where we should do it.
* lib/sqlalchemy/orm/strategy_options.py (line 115): same here....im sure mapper has some string methods but perhaps we can propose something that is cross alias_class, mapper, and obviously when we get an entity the first thing we do is `inspect()` - always - then call upon whatever stringifiy thing we want to do


with this one, perhaps a method like `entity_str()`.   `__str__()` again seems like too big of a catch-all.

can we also look in orm/util.py and/or orm/base.py and see if there are other mapper stringify methods we should be consolidating.
* lib/sqlalchemy/orm/strategy_options.py (line 1928): like the other string methods above, `__str__()` these days feels too all-encompassing.  perhaps `to_option_method_string()` something like that.

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/strategy_options.py (line 115): to be clear:

this is not how i want it.  i want methods on PathRegistry, Mapper, AliasedInsp.  perhaps a single method `.path_string()` common to all of them.     also look around in orm/base.py like i said.  look at methods like `entity_str()`, the looking at the class inside of `state_class_str()`.  this is an explosion of funcitons currently, all stringiying differently.   I'd like this to be much more organized into a single chain of methods and a few entry functions.

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/base.py (line 399): this cast is wrong
* lib/sqlalchemy/orm/strategy_options.py (line 103): Done
* lib/sqlalchemy/orm/strategy_options.py (line 115): Done
* lib/sqlalchemy/orm/strategy_options.py (line 1928): Done

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/base.py (line 399): Done
* lib/sqlalchemy/orm/interfaces.py (line 766): dont need this, per typing these are non-None
* lib/sqlalchemy/orm/path_registry.py (line 485): OK I was wrong, this is way too complicated when we just have `self.path`

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

* lib/sqlalchemy/orm/path_registry.py (line 369): since this is in the loop now we should just make a single conditional for what we append

sqla-tester commented on lib/sqlalchemy/orm/strategy_options.py:
```diff
@@ -1865,9 +1865,25 @@ def _reconcile(
+        if self.local_opts:
+            args["local_opts"] = dict(self.local_opts)
+
+        args_str = ", ".join(f"{k}={v}" for k, v in args.items())
```
**Michael Bayer** (zzzeek) wrote:

should be private

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/6828 has been **merged**. Congratulations! :)
